clang/include/clang/AST/Type.h

```diff
@@ -3173,41 +3174,46 @@ class ArrayType : public Type, public llvm::FoldingSetNode {
     return T->getTypeClass() == ConstantArray ||
            T->getTypeClass() == VariableArray ||
            T->getTypeClass() == IncompleteArray ||
-           T->getTypeClass() == DependentSizedArray;
+           T->getTypeClass() == DependentSizedArray ||
+           T->getTypeClass() == ArrayParameter;
   }
 };
 
 /// Represents the canonical version of C arrays with a specified constant size.
 /// For example, the canonical type for 'int A[4 + 4*100]' is a
 /// ConstantArrayType where the element type is 'int' and the size is 404.
-class ConstantArrayType final
-    : public ArrayType,
-      private llvm::TrailingObjects<ConstantArrayType, const Expr *> {
```

Comment:
The size expression is usually not part of a constant array type. Normally, all we want as part of the type node is the bound, and the size expression if needed can be found on the `ArrayTypeLoc`. The rare special case being handled here is that if the array bound is instantiation-dependent but not dependent, then we need to preserve the expression in the type (not just in the `TypeLoc`) because it affects (for example) whether two template declarations involving such a type are considered to declare the same template. Because this is a very rare case, and `ConstantArrayType`s can be relatively common (eg, as the types of string literals), it seemed preferable to not store the `Expr*` if it's not needed, but in absolute terms it's probably a minor difference in memory usage, so it's probably fine to remove this optimization.

If we want to keep the optimization, perhaps we could replace the `Size` and `SizeExpr` fields with a pointer union that can hold either a <= 63-bit array bound (the common case) or a pointer to a slab-allocated pair of `APInt` and `const Expr*` (in the case where the bound doesn't fit in 63 bits or the size expression is instantiation-dependent).

Reply:
@zygoloid do you think the optimization is important to preserve? I guessed it might be safe to eliminate because other array types store the `SizeExpr` pointer, but if it's important I can make a separate PR to change the implementation as you suggested and we can merge that change first.

Reply:
As a general principle we try to keep our AST nodes small. This change isn't going to make a big difference by itself, but abandoning that principle and making a bunch of changes like this would. I think it's worth putting in a small amount of effort here, but if it doesn't work out nicely then it's fine to remove the optimization. We can find other places to save space.

Reply:
My team is trying to be pedantic about tracking our work, so I've filed #85124 to track implementing @zygoloid's suggestion. I'll try and get a PR up for that in the next day or two then update this PR on top of that change.